File: AlgorithmsPackCPP/417. Pacific Atlantic Water Flow.cpp
class Solution {
    int dr[4]{-1,0,1,0};
    int dc[4]{0,1,0,-1};
    
    struct Cell{
        int r;
        int c;
    };
    
    bool isValid(int r, int c, vector<vector<int>>&heights){
        if(r<0||r>=heights.size())
            return false;
        if(c<0||c>=heights[0].size())
            return false;
        return true;
    }
public:
    void bfs(vector<vector<int>>&heights, vector<vector<bool>>&visited, vector<Cell>&starts){
        queue<Cell>q;
        for(auto&ce:starts){
            q.push(ce);
            visited[ce.r][ce.c] = true;
        }
        for(int level = 0, sz = q.size(); !q.empty(); ++level, sz = q.size()){
            while(sz--){
            int r = q.front().r;
            int c = q.front().c;
            q.pop();
            
            for(int d= 0; d < 4; d++){
                int nr = r+dr[d];
                int nc = c+dc[d];
                if(!isValid(nr,nc,heights)||visited[nr][nc]||heights[nr][nc]<heights[r][c])
                    continue;
                
                visited[nr][nc] = true;
                q.push({nr,nc});
            }
        }
        }
    }
    
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        int rows = heights.size();
        int cols = heights[0].size();
        
        vector<vector<bool>>atlanticGraph(rows,vector<bool>(cols));
        vector<vector<bool>>pacificGraph(rows,vector<bool>(cols));
        
        vector<Cell>atlanticStarts;
        vector<Cell>pacificStarts;
        
        for(int r = 0; r < rows; r++){
            atlanticStarts.push_back({r,cols-1});
            pacificStarts.push_back({r,0});
        }
        
        for(int c = 0; c < cols; c++){
            atlanticStarts.push_back({rows-1,c});
            pacificStarts.push_back({0,c});
        }
        
        bfs(heights,atlanticGraph,atlanticStarts);
        bfs(heights,pacificGraph,pacificStarts);
        
        vector<vector<int>>result;
        
        for(int r = 0; r < rows; r++){
            for(int c = 0; c < cols; c++){
                if(atlanticGraph[r][c] && pacificGraph[r][c]){
                    result.push_back({r,c});
                }
            }
        }
        return result;
    }
};

File: AlgorithmsPackPythonSolutions/417. Pacific Atlantic Water Flow.py
class Solution:
    dr = [-1, 0, 1, 0]
    dc = [0, 1, 0, -1]
    
    class Cell:
        def __init__(self, r, c):
            self.r = r
            self.c = c
    
    def isValid(self, r, c, heights):
        if r < 0 or r >= len(heights):
            return False
        if c < 0 or c >= len(heights[0]):
            return False
        return True
    
    def bfs(self, heights, visited, starts):
        q = deque()
        for ce in starts:
            q.append(ce)
            visited[ce.r][ce.c] = True
        
        while q:
            cur = q.popleft()
            r, c = cur.r, cur.c
            
            for d in range(4):
                nr = r + self.dr[d]
                nc = c + self.dc[d]
                if not self.isValid(nr, nc, heights) or visited[nr][nc] or heights[nr][nc] < heights[r][c]:
                    continue
                
                visited[nr][nc] = True
                q.append(self.Cell(nr, nc))
    
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        if not heights or not heights[0]:
            return []
        
        rows, cols = len(heights), len(heights[0])
        atlanticGraph = [[False] * cols for _ in range(rows)]
        pacificGraph = [[False] * cols for _ in range(rows)]
        atlanticStarts = []
        pacificStarts = []
        
        for r in range(rows):
            atlanticStarts.append(self.Cell(r, cols - 1))
            pacificStarts.append(self.Cell(r, 0))
        
        for c in range(cols):
            atlanticStarts.append(self.Cell(rows - 1, c))
            pacificStarts.append(self.Cell(0, c))
        
        self.bfs(heights, atlanticGraph, atlanticStarts)
        self.bfs(heights, pacificGraph, pacificStarts)
        
        result = []
        for r in range(rows):
            for c in range(cols):
                if atlanticGraph[r][c] and pacificGraph[r][c]:
                    result.append([r, c])
        
        return result
